kvserver: register `MultiRaft` service with DRPC server

shubhamdhama opened the pull request:
Enable MultiRaft service on the DRPC server in addition to gRPC. Controlled
by `rpc.experimental_drpc.enabled` (off by default).

To enable this change, we used the interfaces introduced in
https://github.com/cockroachdb/drpc/pull/2. For example, for server-side
stream implementations, we used `RPCMultiRaft_RaftMessageBatchStream`,
which is the common interface between `DRPCMultiRaft_RaftMessageBatchStream` (for DRPC)
and `MultiRaft_RaftMessageBatchServer ` (for gRPC).

Note: This only registers the service; the client is not updated to use the
DRPC client, so this service will not have any functional effect.

Fixes: #146473
Epic: CRDB-48925
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146926)
<!-- Reviewable:end -->

shubhamdhama commented:
```
benchdiff -b --old lastmerge -c 20 -d 1000x ./pkg/sql/tests -r Sysbench/SQL/3node/oltp_read_write
old:  a085c92 Merge #146891
new:  0b9d361 drpc: don't use adapter due to allocations concern
args: benchdiff "-b" "--old" "lastmerge" "-c" "20" "-d" "1000x" "./pkg/sql/tests" "-r" "Sysbench/SQL/3node/oltp_read_write"

name                                           old time/op    new time/op    delta
Sysbench/SQL/3node/oltp_read_write-24            12.5ms ± 1%    12.4ms ± 1%   -0.89%  (p=0.000 n=20+20)
ParallelSysbench/SQL/3node/oltp_read_write-24    1.02ms ± 8%    1.02ms ± 3%     ~     (p=0.235 n=20+19)

name                                           old errs/op    new errs/op    delta
ParallelSysbench/SQL/3node/oltp_read_write-24      0.01 ±76%      0.01 ±59%  -22.95%  (p=0.034 n=20+18)
Sysbench/SQL/3node/oltp_read_write-24              0.00           0.00          ~     (all equal)

name                                           old alloc/op   new alloc/op   delta
Sysbench/SQL/3node/oltp_read_write-24            1.97MB ± 6%    1.79MB ± 3%   -9.16%  (p=0.000 n=20+20)
ParallelSysbench/SQL/3node/oltp_read_write-24    1.67MB ± 2%    1.63MB ± 1%   -2.81%  (p=0.000 n=19+19)

name                                           old allocs/op  new allocs/op  delta
Sysbench/SQL/3node/oltp_read_write-24             9.26k ± 1%     8.02k ± 1%  -13.48%  (p=0.000 n=19+19)
ParallelSysbench/SQL/3node/oltp_read_write-24     7.34k ± 2%     7.13k ± 2%   -2.97%  (p=0.000 n=20+20)
```

shubhamdhama commented:
- [ ] Add tests

shubhamdhama commented:
Here is the sysbench numbers for using adapter that may create some allocations vs directly using clients (first commit vs second commit of this PR):
```
benchdiff -b --old c504f5271ef4be5b7e71ee719bdf66c212574640 -c 20 -d 1000x ./pkg/sql/tests -r Sysbench/SQL/3node/oltp_read_write
old:  c504f52 drpc: register MultiRaft service with DRPC server
new:  0b9d361 drpc: don't use adapter due to allocations concern
args: benchdiff "-b" "--old" "c504f5271ef4be5b7e71ee719bdf66c212574640" "-c" "20" "-d" "1000x" "./pkg/sql/tests" "-r" "Sysbench/SQL/3node/oltp_read_write"

name                                           old time/op    new time/op    delta
Sysbench/SQL/3node/oltp_read_write-24            12.3ms ± 2%    12.3ms ± 1%   ~     (p=0.301 n=20+20)
ParallelSysbench/SQL/3node/oltp_read_write-24    1.01ms ± 4%    1.00ms ± 3%   ~     (p=0.158 n=20+17)

name                                           old errs/op    new errs/op    delta
Sysbench/SQL/3node/oltp_read_write-24              0.00           0.00        ~     (all equal)
ParallelSysbench/SQL/3node/oltp_read_write-24      0.01 ±71%      0.01 ±79%   ~     (p=0.146 n=20+20)

name                                           old alloc/op   new alloc/op   delta
Sysbench/SQL/3node/oltp_read_write-24            1.80MB ± 6%    1.81MB ± 5%   ~     (p=0.229 n=20+20)
ParallelSysbench/SQL/3node/oltp_read_write-24    1.63MB ± 1%    1.63MB ± 1%   ~     (p=1.000 n=20+17)

name                                           old allocs/op  new allocs/op  delta
Sysbench/SQL/3node/oltp_read_write-24             8.02k ± 2%     8.03k ± 2%   ~     (p=0.533 n=20+20)
ParallelSysbench/SQL/3node/oltp_read_write-24     7.14k ± 1%     7.12k ± 2%   ~     (p=0.583 n=19+20)
```

shubhamdhama committed:
kvserver: register `MultiRaft` service with DRPC server

Enable MultiRaft service on the DRPC server in addition to gRPC. Controlled
by `rpc.experimental_drpc.enabled` (off by default).

Epic: CRDB-48925
Release note: None

cthumuluru-crdb commented on pkg/kv/kvserver/raft_transport.go:
```diff
@@ -375,8 +378,29 @@ func newRaftMessageResponse(
+	return (*drpcRaftTransport)(t)
+}
+
+// RaftMessageBatch proxies the incoming requests to the listening server interface.
```
nit: can you please reword the comment? I don't completely follow what you meant by listening service interface.

shubhamdhama replied:
Oh it's a mindless copy of the original comment for the gRPC implementation (see below).

cthumuluru-crdb commented on pkg/kv/kvserver/raft_transport.go:
```diff
@@ -78,7 +79,7 @@ type RaftMessageResponseStream interface {
 // responses (grpc.serverStream) is not safe for concurrent calls to Send.
 type lockedRaftMessageResponseStream struct {
-	wrapped MultiRaft_RaftMessageBatchServer
+	wrapped RPCMultiRaft_RaftMessageBatchStream
```
nit: can you add a comment that, `RPCMultiRaft_RaftMessageBatchStream` is an adapter for gRPC and DRPC?

shubhamdhama replied:
It's not the adapter. It's the common interface between for gRPC and DRPC,
```
type RPCMultiRaft_RaftMessageBatchStream interface {
	Context() context.Context
	Send(*kvserverpb.RaftMessageResponse) error
	Recv() (*kvserverpb.RaftMessageRequestBatch, error)
	RecvMsg(interface{}) error
}
```

cthumuluru-crdb replied:
Yeah, correct. Adapter usually implements an interface. My comment was mainly about adding some additional details on why we have to replace `MultiRaft_RaftMessageBatchServer` with `RPCMultiRaft_RaftMessageBatchStream`. Like we discussed, it is sufficient to add those details to the commit message and that context will be helpful for the reader.

shubhamdhama replied:
We discussed and the intent of this message was we should be explaining why this is changed. I've edited the PR description.

cthumuluru-crdb reviewed (approved):
lgtm!

shubhamdhama commented:
TFTR!

bors r=cthumuluru-crdb
